code/game/objects/effects/portals.dm

```diff
@@ -143,6 +145,44 @@
 	do_sparks(5, 0, loc)
 	qdel(src)
 
+#define UNSTABLE_TIME_DELAY 2 SECONDS
+
+/obj/effect/portal/hand_tele
+	/// After you touch the portal, it will be unstable with high bad teleport chance, this variable contains time when it will be fine again
+	var/unstable_time = 0
+	/// If this is TRUE, you will not be able to teleport with that portal
+	var/inactive = FALSE
+
+/obj/effect/portal/hand_tele/examine(mob/user, infix, suffix)
+	. = ..()
+	if(unstable_time > world.time)
+		. += "<span class='warning'>[src] is shaking, looks very unstable!</span>"
+
+/obj/effect/portal/hand_tele/can_teleport(atom/movable/M)
+	if(inactive)
+		return FALSE
+	return ..()
+
+/obj/effect/portal/hand_tele/teleport(atom/movable/M)
+	. = ..()
+	adjust_unstable()
+
+/obj/effect/portal/hand_tele/proc/adjust_unstable()
+	unstable_time = world.time + UNSTABLE_TIME_DELAY
+	icon_state = fail_icon
+	failchance = 33
+	inactive = TRUE
+	addtimer(CALLBACK(src, PROC_REF(check_unstable)), UNSTABLE_TIME_DELAY)
+	addtimer(VARSET_CALLBACK(src, inactive, FALSE), 0.5 SECONDS) // after unstable is setted you have 0.5 safe seconds to think if you want to use it
+
+/obj/effect/portal/hand_tele/proc/check_unstable()
+	if(unstable_time - world.time > 0.1 SECONDS)
```

Comment:
Why not just have a boolean that gets set in `adjust_unstable` and reset in `check_unstable()`, and use that when checking if the thing is unstable, instead of checking against a time? Should avoid the race condition you're seeing here.

Reply:
Because `adjust_unstable()` and `check_unstable()` could be called multiple times in the same time. 
Every time portal teleports someone, it calls `adjust_unstable()` which starts 2 seconds timer to check if we should reset unstable in proc `check_unstable()`.
But, if portal is already unstable, it means timer is ticking and in two seconds it will became blue and safe and what happends if we teleport something now? This variable allows me to just add 2 seconds from last teleportation and ignore all other delayed stabilizations

Reply:
Okay, I see what you mean now. I didn't realize that this could get called multiple times like that.

I'm still a little wary on this "check to see that we're just barely being called at the right time" code. It might very well be safe, but it feels like it could invite some weird race conditions. Something else that I think you could do would be to pass the current unstable time to the callback as a parameter, and then in the callback function you could check to see if the portal's `unstable_time` is greater than the value that was passed in as a parameter. If that's the case, return early.

Reply:
i think i did what you want in https://github.com/ParadiseSS13/Paradise/pull/21224/commits/cdc6762945f99af1f36b59e455da248c5a0f62e5